zipkin-query/src/main/scala/com/twitter/zipkin/query/ZipkinQueryController.scala

```diff
@@ -0,0 +1,61 @@
+package com.twitter.zipkin.query
+
+import javax.inject.Inject
+
+import com.twitter.finagle.httpx.Request
+import com.twitter.finagle.tracing.SpanId
+import com.twitter.finatra.http.Controller
+import com.twitter.finatra.http.response.ResponseBuilder
+import com.twitter.finatra.request.{QueryParam, RouteParam}
+import com.twitter.util.{Future, Time}
+import com.twitter.zipkin.json.{JsonSpan, JsonSpanBijection}
+import com.twitter.zipkin.query.adjusters.TimeSkewAdjuster
+import com.twitter.zipkin.storage.{DependencyStore, SpanStore}
+
+class ZipkinQueryController @Inject()(spanStore: SpanStore,
+                                      dependencyStore: DependencyStore,
+                                      queryExtractor: QueryExtractor,
+                                      traceIds: QueryTraceIds,
+                                      response: ResponseBuilder) extends Controller {
+
+  private[this] val EmptyTraces = Future.value(Seq.empty[Seq[JsonSpan]])
+
+  get("/api/v1/services") { request: Request =>
+    spanStore.getAllServiceNames()
+  }
+
+  get("/api/v1/traces") { request: Request =>
+    queryExtractor(request) match {
+      case Some(qr) => traceIds(qr).flatMap(getTraces(_, qr.adjustClockSkew))
+      case None => Future.value(response.badRequest())
+    }
+  }
+
+  get("/api/v1/trace/:id") { request: GetTraceRequest =>
+    getTraces(Seq(request.id.toLong), request.adjust_clock_skew).map(_.headOption)
```

Comment:
Would be nice to separate HTTP API concerns from the business logic into different classes, to allow different implementations of the end-point, possibly with alternative (non-http) protocols / discovery mechanisms.